### ISSUE TITLE
allow tenants to have custom S3 endpoints

### ISSUE THREAD

**Issue by jgraettinger** [estuary/flow#892]
As a user, I want to provide an S3-compatible storage tier under a custom DNS name or IP which will be used for journal fragment storage. I'll provide Estuary with a bespoke access key and secret for usage with my storage service.

### Implementation:

Gazette [already supports](https://github.com/gazette/core/blob/master/broker/fragment/store_s3.go#L28-L34) custom S3 endpoints for fragment storage. It also supports custom AWS profiles. These are specified as arguments of the fragment storage URL of the JournalSpec, like:
  `s3://my-bucket-name/and/prefix/?profile=my-profile&endpoint=my.storage.service`

So, most of the work is in appropriately threading through configuration into the construction of these URLs, and also in defining appropriate profiles that brokers are made aware of.

* We would update the `tenants` table to allow for the optional specification of a custom S3-compatible endpoint and corresponding credentials.
* We will map credentials into a Gazette-compatible AWS configuration profile named _exactly_ as the tenant.
* We will also add a new kind of storage mapping provider `Custom` which is essentially a clone of the S3 or GCS provider (taking a bucket and prefix).
* The control-plane, on seeing `Custom` storage mapping provider, will require that the corresponding tenant has a custom storage endpoint configured, and will thread through fragment storage URLs which include both the custom `endpoint` and `profile` arguments, where `profile` is always exactly the tenant name.

### Discussion:

* For now, we will manually ensure that brokers are made aware of custom credential profiles. We can eventually automate this to make it self-service, but not yet.
* We don't _have_ to introduce a `Custom` storage provider, and instead update the `S3` provider to have this behavior, but won't because:
   * The `Custom` provider makes it very obvious as to the expectations around storage.
   * A separate provider allows for `S3` and `Custom` storage to be mixed within the same tenant.

### Example:

A tenant `MyTenant` with a custom storage endpoint `my.storage.com`, having a storage mapping `MyTenant/foo/` with provider `Custom`, bucket `my-bucket`, and prefix `my-prefix/` is applied to a collection `MyTenant/foo/bar`, and produces a fragment URL like:

  `s3://my-bucket/my-prefix/?profile=MyTenant&endpoint=my.storage.com`

### Security:

We _must_ ensure that S3 credentials are only sent to the correct storage service, to prevent a malicious actor from coercing a storage mapping to send a credential to a honey-pot storage service.

We'll maintain this invariant by coupling the `profile` argument (which identifies and selects a set of credentials to use) with the tenant name (the first component of the collection or task).

In other words, in order for a user to coerce a storage endpoint which sends credentials of tenant `Target` to a storage service, like:

    `s3://my-bucket/?profile=Target&endpoint=my.malicious.storage.service.com`

they must already have full ownership and control (an `admin` capability) to the `Target/` tenant in order to have created the malicious storage mapping in the first place.

Note that having an admin grant to `Target/some/scoped/prefix/` is insufficient, as that such a grant doesn't give any access at all to the corresponding row in the `tenants` table -- only a root grant to `Target/` will do.

**Comment by psFried** [comment]
I wanted to share a security-related detail I've uncovered as part of this. Gazette brokers use the `default` profile for any non-custom storage mappings. This means that, if a user were to create a tenant called `default` with a custom storage endpoint, then we would end up sending our credentials to their untrusted endpoint.

One mitigation would be to add `default` to the `illegal_tenant_names` table, and simply prevent anyone from creating a tenant with such a name. I think that would be completely effective. I don't love how that puts the mitigation so far from the sore spot, though. So I'd like to _also_ add a check in some portion of the code that's explicitly dealing with storage mappings, like the `validation` crate.

**Comment by psFried** [comment]
> - We would update the tenants table to allow for the optional specification of a custom S3-compatible endpoint and corresponding credentials.
> ...
> - The control-plane, on seeing Custom storage mapping provider, will require that the corresponding tenant has a custom storage endpoint configured, and will thread through fragment storage URLs which include both the custom endpoint and profile arguments, where profile is always exactly the tenant name.

As I'm working on this, I'm realizing that we don't actually _need_ to include any custom storage data in the `tenants` table until we actually want to automate passing those credentials to the brokers. The check that a storage mapping specifying a custom endpoint matches a `custom_storage_endpoint` column for the tenant is not required _for security purposes_.

Of course we _will_ want to perform that check once we start allowing users to edit their storage mappings. Otherwise, a user might provide a custom storage endpoint and then be confused as to why things aren't working at runtime. But I don't think there's a necessity to do this now. And my inclination is to defer adding custom storage credentials to the control plane until we're ready to do the work of plumbing them through to the data plane because the desired representation of the credentials would be more clear at that point.

**Comment by jgraettinger** [comment]
I'll confirm whether this is an issue when I'm back, but I think there was a crossed wire here. The endpoint should not be part of the storage mapping, as that introduces an easy foot gun which could inadvertently expose credentials
